xilinx_test/host/RunHBMandBlockCopy.cpp

```diff
@@ -0,0 +1,168 @@
+//#define HLSLIB_SIMULATE_OPENCL
+
+#include "../../include/hlslib/xilinx/OpenCL.h"
+#include <iostream>
+#include <assert.h>
+
+template<typename T>
+bool checkBlockHasValue(const std::array<size_t, 3> blockOffsetSource, const std::array<size_t, 3> copyBlockSize, 
+                        const std::array<size_t, 3> blockSizeSource, const T* source, T checkValue) {
+    size_t sourceSliceJmp = blockSizeSource[0] - copyBlockSize[0] + 
+                            (blockSizeSource[1] - copyBlockSize[1])*blockSizeSource[0];
+    size_t srcindex = blockOffsetSource[0] + blockOffsetSource[1] * blockSizeSource[0]
+                          + blockOffsetSource[2] * blockSizeSource[1] * blockSizeSource[0];
+
+    for(size_t sliceCounter = 0; sliceCounter < copyBlockSize[2]; sliceCounter++) {
+      size_t nextaddsource = 0;
+      for(size_t rowCounter = 0; rowCounter < copyBlockSize[1]; rowCounter++) {
+            srcindex += nextaddsource;
+            for(size_t i = 0; i < copyBlockSize[0]; i++) {
+                if(*(source + srcindex + i) != checkValue) {
+                    return false;
+                }
+            }
+            nextaddsource = blockSizeSource[0];
+      }
+      srcindex += sourceSliceJmp + copyBlockSize[0];
+    }
+
+    return true;
+  }
+
+int main(int argc, char **argv) {
+    std::cout << "Initializing OpenCL context..." << std::endl;
+    hlslib::ocl::Context context("xilinx_u280_xdma_201920_3");
+    std::cout << "Done." << std::endl;
+
+    std::cout << "Initializing memory..." << std::flush;
+    std::array<size_t, 3> buf1Size = {5, 5, 5};
+    std::array<size_t, 3> buf2Size = {3, 3, 3};
+    size_t buf1Elems = buf1Size[0]*buf1Size[1]*buf1Size[2];
+    size_t buf2Elems = buf2Size[0]*buf2Size[1]*buf2Size[2];
+    std::vector<double, hlslib::ocl::AlignedAllocator<double, 4096>> memHostBuf1(buf1Elems);
+    //in sw_emu xilinx seems to just take over the ranges provided to it as device memory (during construction), 
+    //so don't use memDeviceBuf2
+    std::vector<double, hlslib::ocl::AlignedAllocator<double, 4096>> memDeviceBuf2(buf2Elems);
+    std::vector<double, hlslib::ocl::AlignedAllocator<double, 4096>> memHostBuf2(buf2Elems);
+    std::fill(memDeviceBuf2.begin(), memDeviceBuf2.end(), 3.0);
+    std::fill(memHostBuf1.begin(), memHostBuf1.end(), 1.0);
+    std::fill(memHostBuf2.begin(), memHostBuf2.end(), 2.0);
+
+    //A program needs to be loaded because Xilinx does not initialize the buffers otherwise
+    if (argc < 2) {
+        std::cerr << "Required argument: xclbin-file. It can be any xclbin file, since the kernel is not actually"
+        << "executed, but if no program is loaded, vitis will not create the buffers" << std::endl << std::flush;
```

Comment:
I'm not sure this is true, doesn't the kernel have to enable the specific banks you are using? If your shell and kernel don't touch DDR bank 3, I don't think you can copy to DDR bank 3. But I'm not 100% sure